### PR TITLE
restore old documentation references

### LINKED ISSUE #39007: [docs] Urls changed => forum links would become invalid
Originally in https://github.com/pytorch/pytorch/issues/18095#issuecomment-633149307:

@vadimkantorov:
Works: https://pytorch.org/docs/stable/torch.html#torch.flip
Breaks: https://pytorch.org/docs/master/torch.html#torch.flip

@t-vi:
They're now separate pages: https://pytorch.org/docs/master/generated/torch.flip.html#torch.flip
...but it would break many links, including on the forums.

@ezyang:
Yes, this will be a release blocker for next release.

cc @ezyang @gchanan @zou3519 @mattip

### PR DESCRIPTION
**Pull request by mattip**
Fixes gh-39007

We replaced actual content with links to generated content in many places to break the documentation into manageable chunks. This caused references like
```
https://pytorch.org/docs/stable/torch.html#torch.flip
```
to become
```
https://pytorch.org/docs/master/generated/torch.flip.html#torch.flip
```
The textual content that was located at the old reference was replaced with a link to the new reference. This PR adds a `<p id="xxx"/p>` reference next to the link, so that the older references from outside tutorials and forums still work: they will bring the user to the link that they can then follow through to see the actual content.

The way this is done is to monkeypatch the sphinx writer method that produces the link. It is ugly but practical, and in my mind not worse than adding javascript to do the same thing.

### CODE REVIEW

**Comment by ShawnZhong** [comment]
That’s cool! 👍 I didn’t know we can override HTML writer in Sphinx before.

File: docs/source/conf.py [hunk 1]
```diff
@@ -194,6 +194,32 @@ def setup(app):
...
+
+def replace(Klass):
+    old_call = Klass.visit_reference
+
+    def visit_reference(self, node):
+        if 'refuri' in node and 'generated' in node.get('refuri'):
+            ref = node.get('refuri')
+            ref_anchor = ref.split('#')
+            if len(ref_anchor) > 1:
+                self.body.append('<p id="{}"/p>'.format(ref_anchor[1]))
```
Comment (ShawnZhong) [inline docs/source/conf.py:217]: The empty `p` tag might introduce an empty line in some cases: 

| Before | After |
| --- | --- |
| image | image | 
| image | image |

**Comment by ShawnZhong** [comment]
I also noticed that multiple references might be added for the same method, and the browser will choose to get to the first occurrence. 

For example, https://5591729-65600975-gh.circle-artifacts.com/0/docs/torch.html#torch.max links to the middle of `torch.clamp`, which is the first reference of `torch.max` in the page. 

image

File: docs/source/conf.py (hunk 1 above)
Comment (mattip) [reply docs/source/conf.py:217]: The tag should not be emitted here at all. Fixed with more hacks to filter out other references and use only the one that is relevant

Comment (ShawnZhong) [reply docs/source/conf.py]: As a minor side note here, we don’t put `p` after `/` for self closing tags, so it should be something like `<p id="..." />`

Comment (mattip) [reply docs/source/conf.py]: fixed, thanks

**Comment by mattip** [comment]
> we can override HTML writer in Sphinx ...

The pandas theme does much more by using [sphinx events](https://github.com/pandas-dev/pydata-sphinx-theme/blob/master/pydata_sphinx_theme/__init__.py#L176) but this PR is an attempt to keep it simple. If it needs more tweaking, maybe the event handling would be cleaner (or just use the javascript PR gh-39032).

**Review by ShawnZhong — APPROVED** [review]

**Comment by mattip** [comment]
ping @jlin27 this is an alternative to gh-39032 which does something similar in javascript

**Comment by mattip** [comment]
@ShawnZhong is this still needed?

**Comment by jlin27** [comment]
Checked locally and works 
```
file:///Users/jplin/github/pytorch/docs/build/html/torch.html#torch.is_tensor
```
brings users to the correct part of the documentation: 
image

**Review by jlin27 — APPROVED** [review]

**Pull request closed without merging**